server/oidc.go

```diff
@@ -0,0 +1,269 @@
+		params:         params,
+		clientRedirect: EndpointCLIAuthCode,
+	}
+
+	s.router.GET(EndpointAuthCallback, s.webOAuth.HandleOIDCCallback)
+	s.router.GET(EndpointOIDCLogin, s.webOAuth.HandleOIDCLogin)
+
+	s.router.GET(EndpointAuthCLICallback, s.cliOAuth.HandleOIDCCallback)
+	s.router.GET(EndpointOIDCCLILogin, s.cliOAuth.HandleOIDCLogin)
+
+	s.router.GET(EndpointCLIAuthCode, func(ctx *gin.Context) {
+		cookie, err := ctx.Request.Cookie(oktaCookieName)
+		if err != nil {
+			ctx.Writer.Write([]byte(err.Error()))
+			return
+		}
+		ctx.Writer.Write([]byte(cookie.Value))
+	})
+}
+
+// HandleOIDCCallback is the handler function for any callback in OAuth. It will
+// eventually redirect the user to the clientRedirect in the oauthEnv
+// See: https://developer.okta.com/docs/guides/sign-into-web-app-redirect/go/main/
+func (o *oauthEnv) HandleOIDCCallback(c *gin.Context) {
+	session, err := sessionStore.Get(c.Request, oktaCookieName)
+	if err != nil {
+		c.AbortWithError(http.StatusForbidden, err)
+		return
+	}
+
+	// Check the state that was returned in the query string is the same as the above state
+	if c.Query("state") == "" || c.Query("state") != session.Values["oauth_state"] {
+		c.AbortWithError(http.StatusForbidden, fmt.Errorf("the state was not as expected"))
+		return
+	}
+
+	// Make sure the code was provided
+	if c.Query("error") != "" {
+		c.AbortWithError(http.StatusForbidden, fmt.Errorf("authorization server returned an error: %s", c.Query("error")))
+		return
+	}
+
+	// Make sure the code was provided
+	if c.Query("code") == "" {
+		c.AbortWithError(http.StatusForbidden, fmt.Errorf("the code was not returned or is not accessible"))
+		return
+	}
+
+	token, err := o.Exchange(
+		context.Background(),
+		c.Query("code"),
+		oauth2.SetAuthURLParam("code_verifier", session.Values["oauth_code_verifier"].(string)),
+	)
+	if err != nil {
+		c.AbortWithError(http.StatusUnauthorized, err)
+		return
+	}
+
+	// Extract the ID Token from OAuth2 token.
+	rawIDToken, ok := token.Extra("id_token").(string)
+	if !ok {
+		c.AbortWithError(http.StatusUnauthorized, fmt.Errorf("id token missing from OAuth2 token"))
+		return
+	}
+	_, err = o.params.verifyToken(rawIDToken)
+
+	if err != nil {
+		c.AbortWithError(http.StatusForbidden, err)
+		return
+	} else {
+		session.Values["access_token"] = token.AccessToken
+
+		session.Save(c.Request, c.Writer)
+	}
+
+	c.Redirect(http.StatusFound, o.clientRedirect)
+}
+
+// HandleOIDCLogin handles redirecting the user to the Okta login, as well as
+// providing it challenge codes.
+// See: https://developer.okta.com/docs/guides/sign-into-web-app-redirect/go/main/
+func (o *oauthEnv) HandleOIDCLogin(c *gin.Context) {
+	c.Header("Cache-Control", "no-cache") // See https://github.com/okta/samples-golang/issues/20
+
+	session, err := sessionStore.Get(c.Request, oktaCookieName)
+	if err != nil {
+		c.AbortWithError(http.StatusInternalServerError, err)
+		return
+	}
+	// Generate a random state parameter for CSRF security
```

Comment:
Code blocks that need explaining with a comment should be refactored out to well-named methods.